soc: esp32: add support to zephyr-sdk, newlibc and cpp

sylvioalves opened the pull request:
This PR adds changes in all Espressif SoCs in order to allow newlibc and cpp build.

This also updates hal_espressif to enable usage of zephyr-sdk toolchain (after https://github.com/zephyrproject-rtos/sdk-ng/pull/426).

Current toolchain is still working and documentation will only be updated when zephyr-sdk gets proper toolchain in place.

sylvioalves committed:
linker: esp32: changes to enable newlibc and cpp support

This PR adds missing configuration to enable newlibc
and cpp code to run in ESP32 SoC. This isn't enough though.
Toolchain changes are also needed and will come up next.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
soc: esp32: add _gettimeofday_r workaround

Add reentrant _gettimeofday_ call so that build
won't fail. This is only a workaround for now.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
soc: esp32: add toolchain name

Add ESP32 and ESP32S2 toolchain name that is addressed
when integrated with zephyr-sdk.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
soc: esp32: add Espressif HAL config

Current Espressif porting requires standard include as
part of hal implementation. compiler_flags.cmake checks for
variant name to keep those stdinc in build.
Instead of using variant name as check, use this new CONFIG
to make it clear and to allow having toolchain integrated
in zephyr-sdk package.
stdinc dependency in hal_espressif will be worked out and removed
soon.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
soc: esp32: decrease region 1 default pool size

Make the available heap for the 128k bank smaller
so that it frees up space for some samples
as pktqueue and smp. In the context of toolchain,
this enables having twister test to pass in those tests
that requires more memory.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
soc: esp32c3: linker changes to enable newlibc and cpp

Adds linker changes to enable cpp and newlibc code.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
linker: esp32s2: changes to enable newlibc and cpp support

This PR adds missing configuration to enable newlibc
and cpp code to run in ESP32S2 SoC. This isn't enough though.
Toolchain changes are also needed and will come up next.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
soc: esp32s2: add _gettimeofday_r workaround

Add reentrant _gettimeofday_ call so that build
won't fail. This is only a workaround for now.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves commented on cmake/compiler/gcc/compiler_flags.cmake:
```diff
@@ -106,7 +106,7 @@ set_compiler_property(PROPERTY cstd -std=)
 if (NOT CONFIG_NEWLIB_LIBC AND
     NOT COMPILER STREQUAL "xcc" AND
-    NOT ZEPHYR_TOOLCHAIN_VARIANT STREQUAL "espressif" AND
+    NOT CONFIG_SOC_ESPRESSIF AND
```
@stephanosio We have discussed about this before. The necessary work to get `hal_espressif` cleaned up regarding `nostdinc` is complex. I ask you to accept this change as is for now.

stephanosio replied:
This is a very ugly hack that effectively allows code to make use of the newlib headers without linking the newlib library, which may or may not work depending on what you are using.

I am ok with this as a temporary workaround; but, as noted in the commit message, this needs to be fixed at some point by either removing the stdinc dependencies or adding the required stdinc dependency implementations in `hal_espressif`.

sylvioalves replied:
Sure, in fact that hack was there for a long time. I explicit added the workaround in commit message in order to tell that this will be re-worked as you noted.

sylvioalves commented on soc/Kconfig:
```diff
@@ -36,6 +36,9 @@ config SOC_COMPATIBLE_NRF52X
 config SOC_COMPATIBLE_NRF52832
 	bool
 
+config SOC_ESPRESSIF
```
Having this `SOC_ESPRESSIF` config is an option to avoid checking for `SOC_ESP32`, `SOC_ESP32S2` and `SOC_ESP32C3`. As there are both `xtensa` and `riscv` support, I have not found a way to create a common definition for Espressif SoCs. `SOC_FAMILY`  won't work as well due to different platforms. Any other suggestion is welcome.

stephanosio replied:
The issue this is trying to address has more to do with the `hal_espressif` than the SoC itself.

I would suggest adding `Kconfig.espressif` under `modules` with `config HAS_ESPRESSIF_HAL` that is selected by the relevant SoCs.

sylvioalves replied:
Sure, that is a nice approach, thanks for the suggestion.

stephanosio commented:
DNM until the new Zephyr SDK is available in the CI.

sylvioalves commented:
> DNM until the new Zephyr SDK is available in the CI so that this can be fully tested in the CI.

@stephanosio No reason to DNM this. Everything works as is using custom toolchain. Well, newlibc and cpp does not work, but I find it ok to merge this PR.

stephanosio commented:
> @stephanosio No reason to DNM this. Everything works as is using custom toolchain. Well, newlibc and cpp does not work, but I find it ok to merge this PR.

The premise for merging a PR is that it works. The green CI in this case is very misrepresentative of the status quo because it is silently skipping all tests for the `esp32` board due to the `espressif` toolchain being not available in the CI image, and it is not running any tests with the Zephyr SDK either because, obviously, the Zephyr SDK release that supports ESP32 is not available at this time **and, more importantly, you did not specify `- zephyr ` for the toolchains in the board `esp32.yaml`.**

stephanosio reviewed (dismissed):
A few issues that need to be addressed:

1. Need to add the `zephyr` toolchain to the list of the supported toolchains for the relevant boards so that twister builds the ESP32 tests using the Zephyr SDK in the CI.
https://github.com/zephyrproject-rtos/zephyr/blob/adc901aa6a39caffd7971ff99456b26414cd3793/boards/xtensa/esp32/esp32.yaml#L5-L6
https://github.com/zephyrproject-rtos/zephyr/blob/adc901aa6a39caffd7971ff99456b26414cd3793/boards/xtensa/esp32s2_saola/esp32s2_saola.yaml#L5-L6

2. There are two build failures using the Zephyr SDK built in https://github.com/zephyrproject-rtos/sdk-ng/pull/426. These need to be fixed.
    ```
    INFO    -  27/444 esp32                     samples/boards/esp32/wifi_station/sample.board.esp32 FAILED Build failure (build)
    INFO    -  36/444 esp32                     samples/arch/smp/pktqueue/sample.smp.pktqueue      FAILED Build failure (build)
    ```

3. `hal_espressif` displays very misleading warnings during the build process when using the Zephyr SDK. These warnings need to be either disabled or updated to take the Zephyr SDK into account.

```
CMake Warning at /home/stephanos/Dev/zephyrproject/modules/hal/espressif/tools/cmake/crosstool_version_check.cmake:8 (message):
  Xtensa toolchain
  /home/stephanos/Downloads/zephyr-sdk-0.14.0-beta1-4-gf5c0d5b/xtensa-espressif_esp32_zephyr-elf/bin/xtensa-espressif_esp32_zephyr-elf-gcc
  version 10.3.0 is not the supported version 8.4.0.  Check Getting Started
  documentation or proceed at own risk.
Call Stack (most recent call first):
  /home/stephanos/Dev/zephyrproject/modules/hal/espressif/components/esp32/project_include.cmake:30 (gcc_version_check)
  /home/stephanos/Dev/zephyrproject/modules/hal/espressif/tools/cmake/build.cmake:309 (include)
  /home/stephanos/Dev/zephyrproject/modules/hal/espressif/tools/cmake/build.cmake:460 (__build_process_project_includes)
  /home/stephanos/Dev/zephyrproject/modules/hal/espressif/tools/cmake/project.cmake:396 (idf_build_process)
  CMakeLists.txt:41 (project)


CMake Warning at /home/stephanos/Dev/zephyrproject/modules/hal/espressif/tools/cmake/crosstool_version_check.cmake:28 (message):
  Xtensa toolchain
  /home/stephanos/Downloads/zephyr-sdk-0.14.0-beta1-4-gf5c0d5b/xtensa-espressif_esp32_zephyr-elf/bin/xtensa-espressif_esp32_zephyr-elf-gcc
  crosstool-ng version xtensa-espressif_esp32_zephyr-elf-gcc (Zephyr SDK
  0.14.0-beta1-4-gf5c0d5b) 10.3.0

  Copyright (C) 2020 Free Software Foundation, Inc.

  This is free software; see the source for copying conditions.  There is NO

  warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.



   doesn't match supported version esp-2020r3. Check Getting Started documentation or proceed at own risk.
Call Stack (most recent call first):
  /home/stephanos/Dev/zephyrproject/modules/hal/espressif/components/esp32/project_include.cmake:31 (crosstool_version_check)
  /home/stephanos/Dev/zephyrproject/modules/hal/espressif/tools/cmake/build.cmake:309 (include)
  /home/stephanos/Dev/zephyrproject/modules/hal/espressif/tools/cmake/build.cmake:460 (__build_process_project_includes)
  /home/stephanos/Dev/zephyrproject/modules/hal/espressif/tools/cmake/project.cmake:396 (idf_build_process)
  CMakeLists.txt:41 (project)
```

stephanosio commented:
@sylvioalves Also what about `esp32c3` (RISC-V)? Can `riscv64-zephyr-elf` toolchain from the Zephyr SDK be used to build the binaries for it?

laurenmurphyx64 commented:
> A few issues that need to be addressed:
> 
> 1. Need to add the `zephyr` toolchain to the list of the supported toolchains for the relevant boards so that twister builds the ESP32 tests using the Zephyr SDK in the CI.
>    https://github.com/zephyrproject-rtos/zephyr/blob/adc901aa6a39caffd7971ff99456b26414cd3793/boards/xtensa/esp32/esp32.yaml#L5-L6
>    
>    https://github.com/zephyrproject-rtos/zephyr/blob/adc901aa6a39caffd7971ff99456b26414cd3793/boards/xtensa/esp32s2_saola/esp32s2_saola.yaml#L5-L6

Plus the ESP32's little brother:

https://github.com/zephyrproject-rtos/zephyr/blob/adc901aa6a39caffd7971ff99456b26414cd3793/boards/xtensa/esp_wrover_kit/esp_wrover_kit.yaml#L5-L6

sylvioalves commented:
> > A few issues that need to be addressed:
> > 
> > 1. Need to add the `zephyr` toolchain to the list of the supported toolchains for the relevant boards so that twister builds the ESP32 tests using the Zephyr SDK in the CI.
> >    https://github.com/zephyrproject-rtos/zephyr/blob/adc901aa6a39caffd7971ff99456b26414cd3793/boards/xtensa/esp32/esp32.yaml#L5-L6
> >    
> >    https://github.com/zephyrproject-rtos/zephyr/blob/adc901aa6a39caffd7971ff99456b26414cd3793/boards/xtensa/esp32s2_saola/esp32s2_saola.yaml#L5-L6
> 
> Plus the ESP32's little brother, I think:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/adc901aa6a39caffd7971ff99456b26414cd3793/boards/xtensa/esp_wrover_kit/esp_wrover_kit.yaml#L5-L6

+ Heltec board.

laurenmurphyx64 commented:
By the way, it might be worth trying out this + the SDK PR with `samples/modules/tflite-micro/hello_world`. It builds and exercises Google's C++ [TensorFlow Lite Micro library](https://github.com/tensorflow/tflite-micro), which probably uses C++ features a lot more extensively than the synchronization sample. I can understand if supporting TFLM is a low priority for Espressif, but theoretically the sample should build and run on all boards that support C++ (it doesn't exercise any hardware). I can confirm it works on all QEMUs except `qemu_nios2` (and `native_posix`)... for some reason... I need to file issues for those... 

(Note that the other TFLM sample, `magic_wand`, works only on one specific board.)

sylvioalves commented:
> By the way, it might be worth trying out this + the SDK PR with `samples/modules/tflite-micro/hello_world`. It builds and exercises Google's C++ [TensorFlow Lite Micro library](https://github.com/tensorflow/tflite-micro), which probably uses C++ features a lot more extensively than the synchronization sample. I can understand if supporting TFLM is a low priority for Espressif, but theoretically the sample should build and run on all boards that support C++ (it doesn't exercise any hardware). I can confirm it works on all QEMUs except `qemu_nios2` (and `native_posix`)... for some reason. I need to file issues for those... and possibly expand the platform_allow for that test...
> 
> (Note that the other TFLM sample, `magic_wand`, works only on one specific board.)

@laurenmurphyx64 tested with ESP32, ESP32S2 and ESP32C3. All working good:  
```c
I (29) boot: ESP-IDF release/v4.3-149-gddfa61bcb0 2nd stage bootloader
I (29) boot: compile time 16:32:45
I (30) boot: chip revision: 1
I (33) boot_comm: chip revision: 1, min. bootloader chip revision: 0
I (40) boot.esp32: SPI Speed      : 40MHz
I (45) boot.esp32: SPI Mode       : DIO
I (50) boot.esp32: SPI Flash Size : 4MB
I (54) boot: Enabling RNG early entropy source...
I (60) boot: Partition Table:
I (63) boot: ## Label            Usage          Type ST Offset   Length
I (70) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (78) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (85) boot:  2 factory          factory app      00 00 00010000 00100000
I (93) boot: End of partition table
I (97) boot_comm: chip revision: 1, min. application chip revision: 0
I (104) esp_image: segment 0: paddr=00010020 vaddr=00000020 size=0001ch (    28) 
I (113) esp_image: segment 1: paddr=00010044 vaddr=3ffb0000 size=00a70h (  2672) load
I (122) esp_image: segment 2: paddr=00010abc vaddr=3ffb0a70 size=00270h (   624) load
I (130) esp_image: segment 3: paddr=00010d34 vaddr=40080000 size=0332ch ( 13100) load
I (143) esp_image: segment 4: paddr=00014068 vaddr=00000000 size=0bfd0h ( 49104) 
I (165) esp_image: segment 5: paddr=00020040 vaddr=3f400040 size=0db58h ( 56152) map
I (186) esp_image: segment 6: paddr=0002dba0 vaddr=00000000 size=02478h (  9336) 
I (190) esp_image: segment 7: paddr=00030020 vaddr=400d0020 size=39c00h (236544) map
I (283) boot: Loaded app from partition at offset 0x10000
I (283) boot: Disabling RNG early entropy source...
*** Booting Zephyr OS build zephyr-v3.0.0-884-g1fdd82f7590b  ***
x_value: 1.0*2^-127, y_value: 1.0*2^-127

x_value: 1.2566366*2^-2, y_value: 1.4910772*2^-2

x_value: 1.2566366*2^-1, y_value: 1.1183078*2^-1

x_value: 1.8849551*2^-1, y_value: 1.677462*2^-1

x_value: 1.2566366*2^0, y_value: 1.9316229*2^-1

x_value: 1.5707957*2^0, y_value: 1.0420598*2^0

x_value: 1.8849551*2^0, y_value: 1.9146791*2^-1

x_value: 1.0995567*2^1, y_value: 1.6435742*2^-1

x_value: 1.2566366*2^1, y_value: 1.0674761*2^-1

x_value: 1.4137159*2^1, y_value: 1.8977352*2^-3

```

sylvioalves commented:
> @sylvioalves Also what about `esp32c3` (RISC-V)? Can `riscv64-zephyr-elf` toolchain from the Zephyr SDK be used to build the binaries for it?

@stephanosio Yes, it works. Twister complains about 2 tests that I had fixed a few days ago and will submit the changes as well.

sylvioalves committed:
samples: tests: exclude esp32 platform

Remove ESP32 and ESP32S2 from not working samples
and tests. Reason for not working is not enough memory
space in RAM to execute it. This can be worked out as next steps.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves commented:
> @sylvioalves Also what about `esp32c3` (RISC-V)? Can `riscv64-zephyr-elf` toolchain from the Zephyr SDK be used to build the binaries for it?

Yes, it can.

sylvioalves commented:
@stephanosio This should be final update. PTAL again.

stephanosio reviewed (commented):
Looks ok to me.

This can be merged when a new Zephyr SDK release with the ESP32 support is available and the CI passes.

stephanosio commented:
> This can be merged when a new Zephyr SDK release with the ESP32 support is available and the CI passes.

Zephyr SDK 0.14.1, which supports ESP32, is now available in the CI.

Rebased to run tests with the new SDK.

stephanosio reviewed (changes requested):
Seeing a lot of build failures in the CI.

Please address them.

sylvioalves commented:
@stephanosio CI seems not able to perform cloning of the `hal_espressif` submodules as from below. Those submodules are required as a few source code and mostly those wifi/ble binaries blobs comes in place. Is there anything we can do to evaluate this?  

```c
-- Found Git: /usr/bin/git (found version "2.25.1") 
-- Initialising new submodule components/bootloader/subproject/components/micro-ecc/micro-ecc...
Cloning into '/__w/zephyr/modules/hal/espressif/components/bootloader/subproject/components/micro-ecc/micro-ecc'...
Submodule path 'components/bootloader/subproject/components/micro-ecc/micro-ecc': checked out 'd037ec89546fad14b5c4d5456c2e23a71e554966'
-- Initialising new submodule components/bt/controller/esp32c3-bt-lib...
Submodule 'components/bt/controller/esp32c3-bt-lib' (https://github.com/espressif/esp32c3-bt-lib) registered for path 'components/bt/controller/esp32c3-bt-lib'
Cloning into '/__w/zephyr/modules/hal/espressif/components/bt/controller/esp32c3-bt-lib'...
Submodule path 'components/bt/controller/esp32c3-bt-lib': checked out '12f00c45ce9c8cf9a9b2e607b4954f12d4191ffb'
-- Initialising new submodule components/bt/controller/lib_esp32...
Submodule 'components/bt/controller/lib_esp32' (https://github.com/espressif/esp32-bt-lib.git) registered for path 'components/bt/controller/lib_esp32'
Cloning into '/__w/zephyr/modules/hal/espressif/components/bt/controller/lib_esp32'...
Submodule path 'components/bt/controller/lib_esp32': checked out 'ec61ca3caa64874e11f39a92654b4160bb5db06e'
-- Initialising new submodule components/esp_wifi/lib...
fatal: Needed a single revision
Unable to find current revision in submodule path 'components/esp_wifi/lib'
CMake Error at /__w/zephyr/modules/hal/espressif/tools/cmake/git_submodules.cmake:48 (message):
  Git submodule init failed for components/esp_wifi/lib
  ```

stephanosio commented:
> CI seems not able to perform cloning of the `hal_espressif` submodules as from below.

This would not be just a problem with the CI, but with the `west.yml` itself.

If I understand correctly, [you need to specify `submodules: true` for the `west update` to fetch the submodules in the module repository](https://docs.zephyrproject.org/3.0.0/guides/west/manifest.html#option-1-boolean).

sylvioalves commented:
> > CI seems not able to perform cloning of the `hal_espressif` submodules as from below.
> 
> This would not be just a problem with the CI, but with the `west.yml` itself.
> 
> If I understand correctly, [you need to specify `submodules: true` for the `west update` to fetch the submodules in the module repository](https://docs.zephyrproject.org/3.0.0/guides/west/manifest.html#option-1-boolean).
> 
> p.s. was it a deliberate decision to not add `submodules: true` in the Zephyr main `west.yml` since that would cause `west update` to take longer for everyone? we could technically add `west espressif update` in the CI workflows, but that is a very ugly solution in my opinion.

@stephanosio Yes, reason that it isn't in there already is due to the time it takes to sync. I have updated this PR adding the submodule to test CI.

What would be the recommendation for the submodules? Perhaps we could add some custom vendor file that needs to be run in CI in a way "no ugly way". What do you think?

stephanosio commented:
> What would be the recommendation for the submodules? Perhaps we could add some custom vendor file that needs to be run in CI in a way "no ugly way".

@mbolivar-nordic What do you think we should do here?

stephanosio commented:
> Can you confirm the failure? It seems not esp32 related:

c5b59282d6bcbcc5eedaf29b386c90c77036440b is the offending commit.

sylvioalves commented:
> > Can you confirm the failure? It seems not esp32 related:
> 
> [c5b5928](https://github.com/zephyrproject-rtos/zephyr/commit/c5b59282d6bcbcc5eedaf29b386c90c77036440b) is the offending commit.

I consider then CI has passed. I'll trigger again after that gets fixed. So we are pending the submodules issue, right?

stephanosio commented:
> So we are pending the submodules issue, right?

Yes.

mbolivar-nordic commented:
So wait, am I hearing that you want to fetch some submodules that contain binary blobs?

sylvioalves commented:
> So wait, am I hearing that you want to fetch some submodules that contain binary blobs?

Until now, `west espressif update` is the documented command necessary to retrieve Wifi/BT submodules/binary blobs.  
However, having Espressif running in CI also requires that. So we want to figured out if there is a solution for that. Let's continue on `#infrastructure` discord channel.

sylvioalves commented on boards/riscv/esp32c3_devkitm/esp32c3_devkitm.yaml:
```diff
@@ -3,4 +3,4 @@ name: ESP32-C3
 arch: riscv
 toolchain:
-  - espressif
+  - zephyr
```
@stephanosio Can we proceed merging this to have new toolchain available and then we handle CI stuff in a dedicated matter? I can restore this entry to `espressif` and modify the documentation of make sure `export ZEPHYR_TOOLCHAIN_VARIANT="zephyr"` is kept. What do you think?

stephanosio replied:
Doing so will make the process of release QA testing rather inconvenient/prone to human errors (unless someone explicitly remembers to restore this patch locally while testing).

Since it is important that we keep building these platforms in the CI so that they do not bit-rot away, I am in favour of adding `west espressif update` in the twister workflow with a FIXME comment and an issue (bug) to keep track of this.

stephanosio replied:
@sylvioalves Let's discuss this in today's dev-review or next week's TSC. If you are available today to join the dev-review meeting, please ping Maureen so she can add it to the agenda.

sylvioalves replied:
I think dev-review was cancelled this week. So only next week's TSC. Anyway, I thought about merging this to enable sdk toolchain but have another PR enabling CI integration.

sylvioalves commented on boards/riscv/esp32c3_devkitm/esp32c3_devkitm.yaml:
```diff
@@ -3,4 +3,4 @@ name: ESP32-C3
 arch: riscv
 toolchain:
-  - espressif
+  - zephyr
```
@stephanosio Can we proceed merging this to have new toolchain available and then we handle CI stuff in a dedicated matter? I can restore this entry to `espressif` and modify the documentation of make sure `export ZEPHYR_TOOLCHAIN_VARIANT="zephyr"` is kept. What do you think?

sylvioalves commented:
Removed last submodules commit and rebased.

sylvioalves committed:
boards: esp32: added ignore tags

Espressif boards cannot have ble and wifi
CI build tests due to binary blobs policies.
This removes refered tests.

west.yml: update hal repository to get updates
that allows building using Zephyr's SDK.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
boards: esp32: update Espressif boards documentation

Zephyr SDK toolchain integration requires all ESP32 family
board's documentation updates.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

stephanosio commented:
@sylvioalves As per https://github.com/zephyrproject-rtos/hal_espressif/pull/111#issuecomment-1119029094, should this PR be marked as "ready for review?"

sylvioalves commented:
> @sylvioalves As per [zephyrproject-rtos/hal_espressif#111 (comment)](https://github.com/zephyrproject-rtos/hal_espressif/pull/111#issuecomment-1119029094), should this PR be marked as "ready for review?"

Now yes, I was waiting CI to pass before making it ready.

sylvioalves commented:
Latest changes are:  

- Updated documentation to meet new toolchain usage.
- Added ignore tags in each ESP32 boards.
- Update west.yml to bring HAL changes related to submodules

stephanosio commented on boards/riscv/esp32c3_devkitm/doc/index.rst:
```diff
@@ -35,104 +35,88 @@ The features include the following:
 
 .. code-block:: console
 
-  # on CMD:
-  set ESPRESSIF_TOOLCHAIN_PATH=%USERPROFILE%\.espressif\tools\zephyr
-  set ZEPHYR_TOOLCHAIN_VARIANT=espressif
+   ***** Booting Zephyr OS vx.x.x-xxx-gxxxxxxxxxxxx *****
+   Hello World! esp32c3_devkitm
 
-  # on PowerShell
-  $env:ESPRESSIF_TOOLCHAIN_PATH="$env:USERPROFILE\.espressif\tools\zephyr"
-  $env:ZEPHYR_TOOLCHAIN_VARIANT="espressif"
+Debugging
+---------
 
-Finally, retrieve required submodules to build this port. This might take a while for the first time:
+As with much custom hardware, the ESP32C3 modules require patches to
+OpenOCD that are not upstreamed. Espressif maintains their own fork of
+the project. The custom OpenOCD can be obtained by running the following extension:
 
 .. code-block:: console
 
-   west espressif update
+   west espressif install
 
-.. note::
+   .. note::
```
This is not correct. Need to remove leading spaces.

sylvioalves replied:
Can this be added into documentation's CI check?

stephanosio replied:
It is already clear if you look at the PR doc build:
https://builds.zephyrproject.io/zephyr/pr/43534/docs/boards/riscv/esp32c3_devkitm/doc/index.html

stephanosio commented on boards/xtensa/esp32/doc/index.rst:
```diff
@@ -45,145 +45,90 @@ System requirements
+   west espressif install
 
-.. note::
+   .. note::
```
same here

stephanosio commented on boards/xtensa/esp32s2_saola/doc/index.rst:
```diff
@@ -32,104 +32,88 @@ The features include the following:
+   west espressif install
 
-.. note::
+   .. note::
```
same here

stephanosio commented on boards/xtensa/esp_wrover_kit/doc/index.rst:
```diff
@@ -442,104 +442,88 @@ Turn the Power Switch to ON, the 5V Power On LED should light up.
+   west espressif install
 
-.. note::
+   .. note::
```
ditto

stephanosio commented on boards/xtensa/heltec_wifi_lora32_v2/doc/index.rst:
```diff
@@ -26,127 +26,87 @@ System requirements
+   west espressif install
 
-.. note::
+   .. note::
```
ditto

stephanosio commented on boards/xtensa/odroid_go/doc/index.rst:
```diff
@@ -75,102 +75,91 @@ features:
-   Hello World! odroid_go
+   west espressif install
+
+   .. note::
```
ditto

stephanosio commented on boards/xtensa/olimex_esp32_evb/doc/index.rst:
```diff
@@ -117,8 +138,40 @@ message in the monitor:
+
+   west espressif install
+
+   .. note::
```
ditto

sylvioalves commented:
Only rebase due to `<zephyr/..`. I think this should be merged before #45517

stephanosio commented on soc/xtensa/esp32/linker.ld:
```diff
@@ -248,17 +260,18 @@ SECTIONS
   #include "linker/common-rom/common-rom-logging.ld"
```
Although not related to this PR, this is incorrect; it should be using `<>` and also `zephyr/`.

A clean-up of these should be done at some point.

sylvioalves replied:
It works as is. Should it?
